# Project: ASSESSING THE SPREAD AND IMPACT OF COVID-19
## Owner: WHO
### Task: Data Inspection & Cleaning
#### Date: 2021-10-17

Our objective in this notebook is to understand the structure of provided data and prepare it for analysis.

In [1]:
# Importing requiste libraries
import pandas as pd

## Project Data

The data for this project was downloaded from the WHO website. Let us load the csv file from our local machine and get a brief look at our data. 

In [2]:
# Loading data from local storage
raw_covid_data = pd.read_csv('./WHO_covid_data.csv',header=0, index_col=False)

In [3]:
# Generating brief report on the structure of our dataframe
raw_covid_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124199 entries, 0 to 124198
Data columns (total 65 columns):
 #   Column                                   Non-Null Count   Dtype  
---  ------                                   --------------   -----  
 0   iso_code                                 124199 non-null  object 
 1   continent                                118567 non-null  object 
 2   location                                 124199 non-null  object 
 3   date                                     124199 non-null  object 
 4   total_cases                              117538 non-null  float64
 5   new_cases                                117534 non-null  float64
 6   new_cases_smoothed                       116520 non-null  float64
 7   total_deaths                             106658 non-null  float64
 8   new_deaths                               106811 non-null  float64
 9   new_deaths_smoothed                      116520 non-null  float64
 10  total_cases_per_million         

In [4]:
# let's look at the shape of our data
raw_covid_data.shape

(124199, 65)

In [5]:
# inspecting the first few rows of data

raw_covid_data.head()

iso_code continent     location        date  total_cases  new_cases  \
0      AFG      Asia  Afghanistan  2020-02-24          5.0        5.0   
1      AFG      Asia  Afghanistan  2020-02-25          5.0        0.0   
2      AFG      Asia  Afghanistan  2020-02-26          5.0        0.0   
3      AFG      Asia  Afghanistan  2020-02-27          5.0        0.0   
4      AFG      Asia  Afghanistan  2020-02-28          5.0        0.0   

   new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  ...  \
0                 NaN           NaN         NaN                  NaN  ...   
1                 NaN           NaN         NaN                  NaN  ...   
2                 NaN           NaN         NaN                  NaN  ...   
3                 NaN           NaN         NaN                  NaN  ...   
4                 NaN           NaN         NaN                  NaN  ...   

   female_smokers  male_smokers  handwashing_facilities  \
0             NaN           NaN                  37.746   
1             NaN           NaN                  37.746   
2             NaN           NaN                  37.746   
3             NaN           NaN                  37.746   
4             NaN           NaN                  37.746   

   hospital_beds_per_thousand  life_expectancy  human_development_index  \
0                         0.5            64.83                    0.511   
1                         0.5            64.83                    0.511   
2                         0.5            64.83                    0.511   
3                         0.5            64.83                    0.511   
4                         0.5            64.83                    0.511   

   excess_mortality_cumulative_absolute  excess_mortality_cumulative  \
0                                   NaN                          NaN   
1                                   NaN                          NaN   
2                                   NaN                          NaN   
3                                   NaN                          NaN   
4                                   NaN                          NaN   

   excess_mortality  excess_mortality_cumulative_per_million  
0               NaN                                      NaN  
1               NaN                                      NaN  
2               NaN                                      NaN  
3               NaN                                      NaN  
4               NaN                                      NaN  

[5 rows x 65 columns]

In [6]:
# inspecting last 5 rows of our data
raw_covid_data.tail()

iso_code continent  location        date  total_cases  new_cases  \
124194      ZWE    Africa  Zimbabwe  2021-10-12     132016.0      141.0   
124195      ZWE    Africa  Zimbabwe  2021-10-13     132108.0       92.0   
124196      ZWE    Africa  Zimbabwe  2021-10-14     132251.0      143.0   
124197      ZWE    Africa  Zimbabwe  2021-10-15     132285.0       34.0   
124198      ZWE    Africa  Zimbabwe  2021-10-16     132333.0       48.0   

        new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  \
124194             115.857        4645.0         2.0                2.571   
124195              96.286        4648.0         3.0                2.571   
124196             104.000        4655.0         7.0                3.429   
124197              82.857        4655.0         0.0                3.000   
124198              81.571        4657.0         2.0                3.000   

        ...  female_smokers  male_smokers  handwashing_facilities  \
124194  ...             1.6          30.7                  36.791   
124195  ...             1.6          30.7                  36.791   
124196  ...             1.6          30.7                  36.791   
124197  ...             1.6          30.7                  36.791   
124198  ...             1.6          30.7                  36.791   

        hospital_beds_per_thousand  life_expectancy  human_development_index  \
124194                         1.7            61.49                    0.571   
124195                         1.7            61.49                    0.571   
124196                         1.7            61.49                    0.571   
124197                         1.7            61.49                    0.571   
124198                         1.7            61.49                    0.571   

        excess_mortality_cumulative_absolute  excess_mortality_cumulative  \
124194                                   NaN                          NaN   
124195                                   NaN                          NaN   
124196                                   NaN                          NaN   
124197                                   NaN                          NaN   
124198                                   NaN                          NaN   

        excess_mortality  excess_mortality_cumulative_per_million  
124194               NaN                                      NaN  
124195               NaN                                      NaN  
124196               NaN                                      NaN  
124197               NaN                                      NaN  
124198               NaN                                      NaN  

[5 rows x 65 columns]

### Data inspection notes
The data consists of 65 columns and 124199 rows. At a glance, we note that many columns have missing values.Additionally, we have columns where the dtype does not match the data that it represents for example the date column currently has of dtype object instead of a datetime.  


### Trimming the data
Before we proceed to data cleaning we shall need to trim the data to include only columns that useful in answering project objectives.  After thorough consultation, the analytics team have narrowed down 15 columns needed for analysis.

In [7]:
# making copy of our data 

covid_data = raw_covid_data.copy()

In [8]:
# selecting columns needed for analysis

columns = ['iso_code','continent', 'location', 'date', 'total_cases', 
           'new_cases','total_deaths', 'new_deaths','population',
           'median_age', 'aged_65_older','aged_70_older', 'gdp_per_capita',
           'life_expectancy', 'human_development_index', ]

In [9]:
#trimming the dataset

covid_data = covid_data[columns] 

In [10]:
# inspecting our trimmed data set
covid_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124199 entries, 0 to 124198
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   iso_code                 124199 non-null  object 
 1   continent                118567 non-null  object 
 2   location                 124199 non-null  object 
 3   date                     124199 non-null  object 
 4   total_cases              117538 non-null  float64
 5   new_cases                117534 non-null  float64
 6   total_deaths             106658 non-null  float64
 7   new_deaths               106811 non-null  float64
 8   population               123325 non-null  float64
 9   median_age               109489 non-null  float64
 10  aged_65_older            108275 non-null  float64
 11  aged_70_older            108890 non-null  float64
 12  gdp_per_capita           110147 non-null  float64
 13  life_expectancy          117800 non-null  float64
 14  huma

### Inspection notes
We now have the data that we need for our analysis. However, as earlier noted, some datatypes inferred by pandas during data loading are not correct: for example, the date column has a dytpe of object. We also have several columns that can take only whole numbers which should be int64 dtype instead of float64. Therefore, the next step is to ensure that our datatypes match the data that it represents. 

In [11]:
# converting the date column from object dtype to datetime 

covid_data['date'] = pd.to_datetime(covid_data['date'])

In [12]:
# converting columns that take whole numbers to int64 dtype

int_columns = ['total_cases', 'new_cases','total_deaths', 
               'new_deaths', 'population']


### Preparation of our data for data type change

Before we can proceed to change our data type from float64 to int64, we need to deal with all NaN values as they will result in a conversion error. Let us inspect the first few rows of data to determine the right approach to dealing with null values.

In [13]:
covid_data.head(15)

iso_code continent     location       date  total_cases  new_cases  \
0       AFG      Asia  Afghanistan 2020-02-24          5.0        5.0   
1       AFG      Asia  Afghanistan 2020-02-25          5.0        0.0   
2       AFG      Asia  Afghanistan 2020-02-26          5.0        0.0   
3       AFG      Asia  Afghanistan 2020-02-27          5.0        0.0   
4       AFG      Asia  Afghanistan 2020-02-28          5.0        0.0   
5       AFG      Asia  Afghanistan 2020-02-29          5.0        0.0   
6       AFG      Asia  Afghanistan 2020-03-01          5.0        0.0   
7       AFG      Asia  Afghanistan 2020-03-02          5.0        0.0   
8       AFG      Asia  Afghanistan 2020-03-03          5.0        0.0   
9       AFG      Asia  Afghanistan 2020-03-04          5.0        0.0   
10      AFG      Asia  Afghanistan 2020-03-05          5.0        0.0   
11      AFG      Asia  Afghanistan 2020-03-06          5.0        0.0   
12      AFG      Asia  Afghanistan 2020-03-07          8.0        3.0   
13      AFG      Asia  Afghanistan 2020-03-08          8.0        0.0   
14      AFG      Asia  Afghanistan 2020-03-09          8.0        0.0   

    total_deaths  new_deaths  population  median_age  aged_65_older  \
0            NaN         NaN  39835428.0        18.6          2.581   
1            NaN         NaN  39835428.0        18.6          2.581   
2            NaN         NaN  39835428.0        18.6          2.581   
3            NaN         NaN  39835428.0        18.6          2.581   
4            NaN         NaN  39835428.0        18.6          2.581   
5            NaN         NaN  39835428.0        18.6          2.581   
6            NaN         NaN  39835428.0        18.6          2.581   
7            NaN         NaN  39835428.0        18.6          2.581   
8            NaN         NaN  39835428.0        18.6          2.581   
9            NaN         NaN  39835428.0        18.6          2.581   
10           NaN         NaN  39835428.0        18.6          2.581   
11           NaN         NaN  39835428.0        18.6          2.581   
12           NaN         NaN  39835428.0        18.6          2.581   
13           NaN         NaN  39835428.0        18.6          2.581   
14           NaN         NaN  39835428.0        18.6          2.581   

    aged_70_older  gdp_per_capita  life_expectancy  human_development_index  
0           1.337        1803.987            64.83                    0.511  
1           1.337        1803.987            64.83                    0.511  
2           1.337        1803.987            64.83                    0.511  
3           1.337        1803.987            64.83                    0.511  
4           1.337        1803.987            64.83                    0.511  
5           1.337        1803.987            64.83                    0.511  
6           1.337        1803.987            64.83                    0.511  
7           1.337        1803.987            64.83                    0.511  
8           1.337        1803.987            64.83                    0.511  
9           1.337        1803.987            64.83                    0.511  
10          1.337        1803.987            64.83                    0.511  
11          1.337        1803.987            64.83                    0.511  
12          1.337        1803.987            64.83                    0.511  
13          1.337        1803.987            64.83                    0.511  
14          1.337        1803.987            64.83                    0.511

### Dealing with Null Values 

We shall replace null values with zeros in the following columns: 'total_cases', 'new_cases','total_deaths', 'new_deaths'. 

**Rationale**: Null values are from a time when data was not available for example this would be the case when there were no cases of the disease in a country. 

In [14]:
# filling null values with zero 
covid_data[int_columns] = covid_data[int_columns].fillna(0)

In [15]:
# confirming changes to our dataframe

covid_data.head()

iso_code continent     location       date  total_cases  new_cases  \
0      AFG      Asia  Afghanistan 2020-02-24          5.0        5.0   
1      AFG      Asia  Afghanistan 2020-02-25          5.0        0.0   
2      AFG      Asia  Afghanistan 2020-02-26          5.0        0.0   
3      AFG      Asia  Afghanistan 2020-02-27          5.0        0.0   
4      AFG      Asia  Afghanistan 2020-02-28          5.0        0.0   

   total_deaths  new_deaths  population  median_age  aged_65_older  \
0           0.0         0.0  39835428.0        18.6          2.581   
1           0.0         0.0  39835428.0        18.6          2.581   
2           0.0         0.0  39835428.0        18.6          2.581   
3           0.0         0.0  39835428.0        18.6          2.581   
4           0.0         0.0  39835428.0        18.6          2.581   

   aged_70_older  gdp_per_capita  life_expectancy  human_development_index  
0          1.337        1803.987            64.83                    0.511  
1          1.337        1803.987            64.83                    0.511  
2          1.337        1803.987            64.83                    0.511  
3          1.337        1803.987            64.83                    0.511  
4          1.337        1803.987            64.83                    0.511

In [16]:
# confirming changes to our dataframe

covid_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124199 entries, 0 to 124198
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   iso_code                 124199 non-null  object        
 1   continent                118567 non-null  object        
 2   location                 124199 non-null  object        
 3   date                     124199 non-null  datetime64[ns]
 4   total_cases              124199 non-null  float64       
 5   new_cases                124199 non-null  float64       
 6   total_deaths             124199 non-null  float64       
 7   new_deaths               124199 non-null  float64       
 8   population               124199 non-null  float64       
 9   median_age               109489 non-null  float64       
 10  aged_65_older            108275 non-null  float64       
 11  aged_70_older            108890 non-null  float64       
 12  gdp_per_capita  

In [17]:
# we can now convert these columns to int64 without running into any conversion errors

covid_data[int_columns] = covid_data[int_columns].astype('int64')

In [18]:
# confirm changes to dataframe

covid_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124199 entries, 0 to 124198
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   iso_code                 124199 non-null  object        
 1   continent                118567 non-null  object        
 2   location                 124199 non-null  object        
 3   date                     124199 non-null  datetime64[ns]
 4   total_cases              124199 non-null  int64         
 5   new_cases                124199 non-null  int64         
 6   total_deaths             124199 non-null  int64         
 7   new_deaths               124199 non-null  int64         
 8   population               124199 non-null  int64         
 9   median_age               109489 non-null  float64       
 10  aged_65_older            108275 non-null  float64       
 11  aged_70_older            108890 non-null  float64       
 12  gdp_per_capita  

### Checking for missing values 
From initial assessments and inspection of data, we observe that 'continent' column is missing some values. Let us take a closer look to identify the pattern of missing data.

In [19]:
covid_data[covid_data['continent'].isnull()]

iso_code continent location       date  total_cases  new_cases  \
601     OWID_AFR       NaN   Africa 2020-02-13            0          0   
602     OWID_AFR       NaN   Africa 2020-02-14            1          1   
603     OWID_AFR       NaN   Africa 2020-02-15            1          0   
604     OWID_AFR       NaN   Africa 2020-02-16            1          0   
605     OWID_AFR       NaN   Africa 2020-02-17            1          0   
...          ...       ...      ...        ...          ...        ...   
122485  OWID_WRL       NaN    World 2021-10-12    238705193     432550   
122486  OWID_WRL       NaN    World 2021-10-13    239167859     462666   
122487  OWID_WRL       NaN    World 2021-10-14    239608139     440280   
122488  OWID_WRL       NaN    World 2021-10-15    240070992     462853   
122489  OWID_WRL       NaN    World 2021-10-16    240402906     331914   

        total_deaths  new_deaths  population  median_age  aged_65_older  \
601                0           0  1373486472         NaN            NaN   
602                0           0  1373486472         NaN            NaN   
603                0           0  1373486472         NaN            NaN   
604                0           0  1373486472         NaN            NaN   
605                0           0  1373486472         NaN            NaN   
...              ...         ...         ...         ...            ...   
122485       4865619        8199  7874965730        30.9          8.696   
122486       4874258        8639  7874965730        30.9          8.696   
122487       4882066        7808  7874965730        30.9          8.696   
122488       4889737        7671  7874965730        30.9          8.696   
122489       4894860        5123  7874965730        30.9          8.696   

        aged_70_older  gdp_per_capita  life_expectancy  \
601               NaN             NaN              NaN   
602               NaN             NaN              NaN   
603               NaN             NaN              NaN   
604               NaN             NaN              NaN   
605               NaN             NaN              NaN   
...               ...             ...              ...   
122485          5.355       15469.207            72.58   
122486          5.355       15469.207            72.58   
122487          5.355       15469.207            72.58   
122488          5.355       15469.207            72.58   
122489          5.355       15469.207            72.58   

        human_development_index  
601                         NaN  
602                         NaN  
603                         NaN  
604                         NaN  
605                         NaN  
...                         ...  
122485                    0.737  
122486                    0.737  
122487                    0.737  
122488                    0.737  
122489                    0.737  

[5632 rows x 15 columns]

In [20]:
covid_data[covid_data['continent'].isnull()]['location'].unique()

array(['Africa', 'Asia', 'Europe', 'European Union', 'International',
       'North America', 'Oceania', 'South America', 'World'], dtype=object)

This reveals that when the location is the name of a continent, for example 'africa' or 'Asia', the continent column will be null. To remedy this issue, It would be more appropriate to split our data into two sets, one containing only country/location data and another containing only continent data. 

This is a case where aggregated data has been mixed with original data. 

In [21]:
# seperate location and continent data

location_covid_data = covid_data[covid_data['continent'].notnull()]
continent_covid_data = covid_data[covid_data['continent'].isnull()]

### Focus on location data

Since continent data is just a summary of location data, from this point on we shall focus on cleaning location_covid_data. Later on we can aggregate this data to get continent data.

In [22]:
location_covid_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 118567 entries, 0 to 124198
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   iso_code                 118567 non-null  object        
 1   continent                118567 non-null  object        
 2   location                 118567 non-null  object        
 3   date                     118567 non-null  datetime64[ns]
 4   total_cases              118567 non-null  int64         
 5   new_cases                118567 non-null  int64         
 6   total_deaths             118567 non-null  int64         
 7   new_deaths               118567 non-null  int64         
 8   population               118567 non-null  int64         
 9   median_age               108855 non-null  float64       
 10  aged_65_older            107641 non-null  float64       
 11  aged_70_older            108256 non-null  float64       
 12  gdp_per_capita  

In [23]:
# let us inspect null columns
location_covid_data[location_covid_data['median_age'].isnull()]['location'].unique()

array(['Andorra', 'Anguilla', 'Bermuda',
       'Bonaire Sint Eustatius and Saba', 'British Virgin Islands',
       'Cayman Islands', 'Cook Islands', 'Dominica', 'Faeroe Islands',
       'Falkland Islands', 'Gibraltar', 'Greenland', 'Guernsey',
       'Isle of Man', 'Jersey', 'Kosovo', 'Liechtenstein',
       'Marshall Islands', 'Monaco', 'Montserrat', 'Nauru', 'Niue',
       'Northern Cyprus', 'Palau', 'Pitcairn', 'Saint Helena',
       'Saint Kitts and Nevis', 'San Marino', 'Sint Maarten (Dutch part)',
       'Tokelau', 'Turks and Caicos Islands', 'Tuvalu', 'Vatican',
       'Wallis and Futuna'], dtype=object)

In [24]:
location_covid_data[location_covid_data['gdp_per_capita'].isnull()]['location'].unique()

array(['Andorra', 'Anguilla', 'Bonaire Sint Eustatius and Saba',
       'British Virgin Islands', 'Cook Islands', 'Cuba', 'Curacao',
       'Faeroe Islands', 'Falkland Islands', 'French Polynesia',
       'Gibraltar', 'Greenland', 'Guernsey', 'Isle of Man', 'Jersey',
       'Liechtenstein', 'Monaco', 'Montserrat', 'New Caledonia', 'Niue',
       'Northern Cyprus', 'Pitcairn', 'Saint Helena', 'Somalia', 'Syria',
       'Taiwan', 'Tokelau', 'Turks and Caicos Islands', 'Vatican',
       'Wallis and Futuna'], dtype=object)

## Inspection Notes
We note that countries with missing data are mostly small countries and islands in the pacific. At this point we could simply discard these data points as they are likely to have very little effect on the final analysis. However, the nature of this project that every country be considered in the analysis. 

## Managing missing values 

Previously, we opted to fill missing values with zeros. For the remaining variables/columns, this is not an option as it would result in incorrect data and ultimately wrong conclusions. The last inspection of our data revealed data gaps in the median_age, aged_65_older, aged_70_older, gdp_per_capita, life_expectancy, and human_development_index columns. 

### Using Proxy Data

We will make to following assumption:
- Countries in the same continent have nearly similar values for mean_age, aged_65_older, aged_70_older, gdp_per_capita, life_expectancy, and human_development_index columns

We will use continent averages in place of the missing data. Example: The average median_age of countries in north-america is used if a country in north america is missing the median_age value.This is repeated for the variables/columns aged_65_older, aged_70_older, gdp_per_capita, life_expectancy, and human_development_index.

In [25]:
# list of columns with null values

columns_with_nulls = ['median_age', 'aged_65_older', 'aged_70_older', 'gdp_per_capita', 
                      'life_expectancy', 'human_development_index']

# iterating through list of columns with nulls and replacing null values with continent averages

for column in columns_with_nulls:
    # create list containing index of null values
    index_of_nulls = location_covid_data[location_covid_data[column].isnull()].index.tolist()

    # Using non-null values to calculate variable average for each continent

    variable_mean_per_continent = location_covid_data[location_covid_data[column].notnull()].groupby('continent')[column].mean().round(3)

    # creating a dictionary from mean_median_age_per_continent

    variable_mean_per_continent_dict = variable_mean_per_continent.to_dict()

    # replacing null values in dataset with continent_mean_median_age 
    # we use the dictionary to map mean_median_age data to a datapoint in a continent 

    location_covid_data.loc[index_of_nulls, column] = location_covid_data.loc[index_of_nulls, 'continent'].map(variable_mean_per_continent_dict)


C:\Users\Kipngeno\.conda\envs\note\lib\site-packages\pandas\core\indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


In [26]:
# checking that the null replacement was done correctly
# the last loop stored index for null values in the human_development index column
location_covid_data.loc[index_of_nulls, ['continent', 'human_development_index']]

continent  human_development_index
3583    North America                    0.757
3584    North America                    0.757
3585    North America                    0.757
3586    North America                    0.757
3587    North America                    0.757
...               ...                      ...
121851        Oceania                    0.731
121852        Oceania                    0.731
121853        Oceania                    0.731
121854        Oceania                    0.731
121855        Oceania                    0.731

[9350 rows x 2 columns]

In [27]:
# We can cross reference the above with our dictionary to confirm the replacement
variable_mean_per_continent_dict

{'Africa': 0.562,
 'Asia': 0.742,
 'Europe': 0.88,
 'North America': 0.757,
 'Oceania': 0.731,
 'South America': 0.764}

In [28]:
# Let's inspect the data to confirm that all the null values have been replaced

location_covid_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 118567 entries, 0 to 124198
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   iso_code                 118567 non-null  object        
 1   continent                118567 non-null  object        
 2   location                 118567 non-null  object        
 3   date                     118567 non-null  datetime64[ns]
 4   total_cases              118567 non-null  int64         
 5   new_cases                118567 non-null  int64         
 6   total_deaths             118567 non-null  int64         
 7   new_deaths               118567 non-null  int64         
 8   population               118567 non-null  int64         
 9   median_age               118567 non-null  float64       
 10  aged_65_older            118567 non-null  float64       
 11  aged_70_older            118567 non-null  float64       
 12  gdp_per_capita  

In [29]:
location_covid_data.head()

iso_code continent     location       date  total_cases  new_cases  \
0      AFG      Asia  Afghanistan 2020-02-24            5          5   
1      AFG      Asia  Afghanistan 2020-02-25            5          0   
2      AFG      Asia  Afghanistan 2020-02-26            5          0   
3      AFG      Asia  Afghanistan 2020-02-27            5          0   
4      AFG      Asia  Afghanistan 2020-02-28            5          0   

   total_deaths  new_deaths  population  median_age  aged_65_older  \
0             0           0    39835428        18.6          2.581   
1             0           0    39835428        18.6          2.581   
2             0           0    39835428        18.6          2.581   
3             0           0    39835428        18.6          2.581   
4             0           0    39835428        18.6          2.581   

   aged_70_older  gdp_per_capita  life_expectancy  human_development_index  
0          1.337        1803.987            64.83                    0.511  
1          1.337        1803.987            64.83                    0.511  
2          1.337        1803.987            64.83                    0.511  
3          1.337        1803.987            64.83                    0.511  
4          1.337        1803.987            64.83                    0.511

## Inspection notes

Our checks and inspection reveal that all the missing data has been filled in the correctly. While filling missing values we used the round method to round of averages to three decimal places. For the median age and life_expectancy columns, we need to make a few adjustments because non-null data has one and two decimal places respectively.

In [32]:
# rounding off data in median_age column to 1 decimal place. 
location_covid_data.loc[:, 'median_age'] = location_covid_data.loc[:, 'median_age'].round(1)

In [31]:
# rounding off data in life_expectancy column to 2 decimal places
location_covid_data.loc[:, 'life_expectancy'] = location_covid_data.loc[:, 'life_expectancy'].round(2)

## Summary

We have now completed our data cleaning process. Our dataset is complete and consistent across all columns. We can confidently use this data for data analysis. Below is an outline of data cleaning process:
1. Trim data to include only columns needed in analysis.
2. Change data types to match data that it represents
3. Fill in missing values
4. Check for consistency

In [1]:
# export csv file
location_covid_data.to_csv('./location_covid_data.csv', index=False)

Object `location_covid_data.to_csv` not found.
